File: 2023/domaci/38-souborovy-system/volfmatous/main.py
from typing import List


def divide_path(path: str) -> List[str]:
    path_split = path.split("/")
    path_is_absolute = path.startswith("/")

    path_simplified = []
    i = 1 if path_is_absolute else 0
    while i < len(path_split) - 1:
        segment = path_split[i]

        if path_is_absolute and segment == "..":
            if len(path_simplified) < 1:
                i += 1
                continue

            path_simplified.pop()
        else:
            path_simplified.append(segment)

        i += 1

    return path_simplified


if __name__ == "__main__":
    path1 = input()
    path2 = input()

    path2_is_absolute = path2.startswith("/")

    path1_segments = divide_path(path1)
    path2_segments = divide_path(path2)

    result = ""

    if path2_is_absolute:
        longer_path_length = max(len(path1_segments), len(path2_segments))
        i = 0
        while i < longer_path_length:
            if i < len(path1_segments) and i < len(path2_segments) and path1_segments[i] == path2_segments[i]:
                i += 1
                continue

            j = i
            while j < len(path1_segments):
                result += "../"
                j += 1

            j = i
            while j < len(path2_segments):
                result += path2_segments[j] + "/"
                j += 1

            break
    else:
        for segment in path2_segments:
            if segment == "..":
                if len(path1_segments) < 1:
                    continue

                path1_segments.pop()
            else:
                path1_segments.append(segment)

        result = "/"
        for segment in path1_segments:
            result += segment + "/"

    print(result)

File: 2023/domaci/38-souborovy-system/volfmatous/main.js
const readline = require("readline");
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: false,
});

let linesRead = 0;
const lines = [];

async function readLine() {
    if (lines.length < 1) {
        for await (const line of rl) {
            lines.push(line);
        }
    }

    linesRead++;
    return lines[linesRead - 1];
}

main().then();

async function main() {
    const path1 = await readLine();
    const path2 = await readLine();

    const path2IsAbsolute = path2.startsWith("/");

    const path1Segments = dividePath(path1);
    const path2Segments = dividePath(path2);

    let result = "";

    if (path2IsAbsolute) {
        const longerPathLength = Math.max(path1Segments.length, path2Segments.length);
        for (let i = 0; i < longerPathLength; i++) {
            if (i < path1Segments.length && i < path2Segments.length && path1Segments[i] === path2Segments[i]) {
                continue;
            }

            for (let j = i; j < path1Segments.length; j++) {
                result += "../";
            }

            for (let j = i; j < path2Segments.length; j++) {
                result += path2Segments[j] + "/";
            }

            break;
        }
    } else {
        for (const segment of path2Segments) {
            if (segment === "..") {
                if (path1Segments.length < 1) {
                    continue;
                }

                path1Segments.pop();
            } else {
                path1Segments.push(segment);
            }
        }

        result = "/";
        for (const segment of path1Segments) {
            result += segment + "/";
        }
    }

    console.log(result);
}

function dividePath(path) {
    const pathSplit = path.split("/");
    const pathIsAbsolute = path.startsWith("/");

    const pathSimplified = [];
    for (let i = pathIsAbsolute ? 1 : 0; i < pathSplit.length - 1; i++) {
        const segment = pathSplit[i];

        if (pathIsAbsolute && segment === "..") {
            if (pathSimplified.length < 1) {
                continue;
            }

            pathSimplified.pop();
        } else {
            pathSimplified.push(segment);
        }
    }

    return pathSimplified;
}
